{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMeYIPbhkK7Nzbklfavz8J2",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/codebuzer/RNA-seq-transcript-coverage-/blob/main/GeneCoverage.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "jfYpEUqIP90p"
      },
      "outputs": [],
      "source": [
        "import pysam\n",
        "import time\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import subprocess\n",
        "import pandas as pd\n",
        "import time\n",
        "import os\n",
        "\n",
        "df = pd.read_csv(\"gtffile.csv\", index_col=0, low_memory=False)\n",
        "\n",
        "\n",
        "def calculate_gene_coverage(bam_file, df):\n",
        "    # indexing of bam file\n",
        "    start_index = time.time()\n",
        "    subprocess.run([\"samtools\", \"index\", bam_file])\n",
        "    end_index = time.time()\n",
        "    print('Indexing Done For')\n",
        "    print('Indexing takes', end_index - start_index, 'secs')\n",
        "    start_align = time.time()\n",
        "    # readoing a bam file\n",
        "    bam = pysam.AlignmentFile(bam_file, \"rb\")\n",
        "    end_align = time.time()\n",
        "\n",
        "    print('Bam alignment take:', end_align - start_align, 'secs')\n",
        "\n",
        "    coverage_data = []\n",
        "\n",
        "    for _, gene in df.iterrows():\n",
        "        chrom = gene['Region']\n",
        "        start = gene['Start']\n",
        "        end = gene['End']\n",
        "        Region_name = gene['Region_name']\n",
        "        gene_id_name = gene['Description'].split(\";\")[0].split(\" \")[1].replace('\"', '')\n",
        "\n",
        "        region = end - start + 1\n",
        "        start_count = time.time()\n",
        "        # counting the coverage\n",
        "        coverage = bam.count_coverage(chrom, start, end)\n",
        "\n",
        "        # coverage contribuition of each base\n",
        "        coverage_A = coverage[0]\n",
        "        coverage_C = coverage[1]\n",
        "        coverage_G = coverage[2]\n",
        "        coverage_T = coverage[3]\n",
        "\n",
        "        # number of mapped reads of each base\n",
        "        num_reads_mapped_A = sum(coverage_A)\n",
        "        num_reads_mapped_C = sum(coverage_C)\n",
        "        num_reads_mapped_G = sum(coverage_G)\n",
        "        num_reads_mapped_T = sum(coverage_T)\n",
        "\n",
        "        # contraint to return only those coverage which are validated by atleast 10 reads\n",
        "        covbases_A_1 = sum(sum(depth >= 1 for depth in depth_array) for depth_array in [list(coverage_A)])\n",
        "        covbases_A_5 = sum(sum(depth >= 5 for depth in depth_array) for depth_array in [list(coverage_A)])\n",
        "        covbases_A_10 = sum(sum(depth >= 10 for depth in depth_array) for depth_array in [list(coverage_A)])\n",
        "        covbases_A_15 = sum(sum(depth >= 15 for depth in depth_array) for depth_array in [list(coverage_A)])\n",
        "        covbases_A_20 = sum(sum(depth >= 20 for depth in depth_array) for depth_array in [list(coverage_A)])\n",
        "\n",
        "        covbases_C_1 = sum(sum(depth >= 1 for depth in depth_array) for depth_array in [list(coverage_C)])\n",
        "        covbases_C_5 = sum(sum(depth >= 5 for depth in depth_array) for depth_array in [list(coverage_C)])\n",
        "        covbases_C_10 = sum(sum(depth >= 10 for depth in depth_array) for depth_array in [list(coverage_C)])\n",
        "        covbases_C_15 = sum(sum(depth >= 15 for depth in depth_array) for depth_array in [list(coverage_C)])\n",
        "        covbases_C_20 = sum(sum(depth >= 20 for depth in depth_array) for depth_array in [list(coverage_C)])\n",
        "\n",
        "        covbases_G_1 = sum(sum(depth >= 1 for depth in depth_array) for depth_array in [list(coverage_G)])\n",
        "        covbases_G_5 = sum(sum(depth >= 5 for depth in depth_array) for depth_array in [list(coverage_G)])\n",
        "        covbases_G_10 = sum(sum(depth >= 10 for depth in depth_array) for depth_array in [list(coverage_G)])\n",
        "        covbases_G_15 = sum(sum(depth >= 15 for depth in depth_array) for depth_array in [list(coverage_G)])\n",
        "        covbases_G_20 = sum(sum(depth >= 20 for depth in depth_array) for depth_array in [list(coverage_G)])\n",
        "\n",
        "        covbases_T_1 = sum(sum(depth >= 1 for depth in depth_array) for depth_array in [list(coverage_T)])\n",
        "        covbases_T_5 = sum(sum(depth >= 5 for depth in depth_array) for depth_array in [list(coverage_T)])\n",
        "        covbases_T_10 = sum(sum(depth >= 10 for depth in depth_array) for depth_array in [list(coverage_T)])\n",
        "        covbases_T_15 = sum(sum(depth >= 15 for depth in depth_array) for depth_array in [list(coverage_T)])\n",
        "        covbases_T_20 = sum(sum(depth >= 20 for depth in depth_array) for depth_array in [list(coverage_T)])\n",
        "\n",
        "        # percentage contribuition of each base in the coverage of that particular region\n",
        "        covpercent_A_1 = (covbases_A_1 / region) * 100\n",
        "        covpercent_A_5 = (covbases_A_5 / region) * 100\n",
        "        covpercent_A_10 = (covbases_A_10 / region) * 100\n",
        "        covpercent_A_15 = (covbases_A_15 / region) * 100\n",
        "        covpercent_A_20 = (covbases_A_20 / region) * 100\n",
        "\n",
        "        covpercent_C_1 = (covbases_C_1 / region) * 100\n",
        "        covpercent_C_5 = (covbases_C_5 / region) * 100\n",
        "        covpercent_C_10 = (covbases_C_10 / region) * 100\n",
        "        covpercent_C_15 = (covbases_C_15 / region) * 100\n",
        "        covpercent_C_20 = (covbases_C_20 / region) * 100\n",
        "\n",
        "        covpercent_G_1 = (covbases_G_1 / region) * 100\n",
        "        covpercent_G_5 = (covbases_G_5 / region) * 100\n",
        "        covpercent_G_10 = (covbases_G_10 / region) * 100\n",
        "        covpercent_G_15 = (covbases_G_15 / region) * 100\n",
        "        covpercent_G_20 = (covbases_G_20 / region) * 100\n",
        "\n",
        "        covpercent_T_1 = (covbases_T_1 / region) * 100\n",
        "        covpercent_T_5 = (covbases_T_5 / region) * 100\n",
        "        covpercent_T_10 = (covbases_T_10 / region) * 100\n",
        "        covpercent_T_15 = (covbases_T_15 / region) * 100\n",
        "        covpercent_T_20 = (covbases_T_20 / region) * 100\n",
        "\n",
        "        # Overall all coverage contribuited by all the bases\n",
        "        covbases_1 = sum(sum(depth >= 1 for depth in depth_array) for depth_array in coverage)\n",
        "        covbases_5 = sum(sum(depth >= 5 for depth in depth_array) for depth_array in coverage)\n",
        "        covbases_10 = sum(sum(depth >= 10 for depth in depth_array) for depth_array in coverage)\n",
        "        covbases_15 = sum(sum(depth >= 15 for depth in depth_array) for depth_array in coverage)\n",
        "        covbases_20 = sum(sum(depth >= 20 for depth in depth_array) for depth_array in coverage)\n",
        "        covbases_25 = sum(sum(depth >= 25 for depth in depth_array) for depth_array in coverage)\n",
        "        covbases_30 = sum(sum(depth >= 30 for depth in depth_array) for depth_array in coverage)\n",
        "\n",
        "        # coverage percent of all the bases\n",
        "        covpercent_1 = (covbases_1 / region) * 100\n",
        "        covpercent_5 = (covbases_5 / region) * 100\n",
        "        covpercent_10 = (covbases_10 / region) * 100\n",
        "        covpercent_15 = (covbases_15 / region) * 100\n",
        "        covpercent_20 = (covbases_20 / region) * 100\n",
        "        covpercent_25 = (covbases_25 / region) * 100\n",
        "        covpercent_30 = (covbases_30 / region) * 100\n",
        "\n",
        "        # Adjust covbases and covpercent if necessary\n",
        "        covbases_1 = np.where(covbases_1 > region, region, covbases_1)\n",
        "        covbases_5 = np.where(covbases_5 > region, region, covbases_5)\n",
        "        covbases_10 = np.where(covbases_10 > region, region, covbases_10)\n",
        "        covbases_15 = np.where(covbases_15 > region, region, covbases_15)\n",
        "        covbases_20 = np.where(covbases_20 > region, region, covbases_20)\n",
        "        covbases_25 = np.where(covbases_25 > region, region, covbases_25)\n",
        "\n",
        "        covpercent_1 = np.where(covpercent_1 > 100, 100.0, covpercent_1)\n",
        "        covpercent_5 = np.where(covpercent_5 > 100, 100.0, covpercent_5)\n",
        "        covpercent_10 = np.where(covpercent_10 > 100, 100.0, covpercent_10)\n",
        "        covpercent_15 = np.where(covpercent_15 > 100, 100.0, covpercent_15)\n",
        "        covpercent_20 = np.where(covpercent_20 > 100, 100.0, covpercent_20)\n",
        "        covpercent_25 = np.where(covpercent_25 > 100, 100.0, covpercent_25)\n",
        "\n",
        "        # please paste here the coverage_data.append\n",
        "\n",
        "        coverage_data.append((Region_name, gene_id_name, chrom, start, end,\n",
        "                              covpercent_1, covpercent_5, covpercent_10,\n",
        "                              covpercent_15, covpercent_20, covpercent_25,\n",
        "                              covbases_1, covbases_5, covbases_10, covbases_15,\n",
        "                              covbases_20, covbases_25, covpercent_G_1, covpercent_G_5,\n",
        "                              covpercent_G_10, covpercent_G_15, covpercent_G_20,\n",
        "                              covpercent_C_1, covpercent_C_5, covpercent_C_10,\n",
        "                              covpercent_C_15, covpercent_C_20,\n",
        "                              covpercent_T_1, covpercent_T_5, covpercent_T_10, covpercent_T_15,\n",
        "                              covpercent_T_20, covpercent_A_1, covpercent_A_5, covpercent_A_10,\n",
        "                              covpercent_A_15, covpercent_A_20,\n",
        "                              covbases_A_1, covbases_A_5, covbases_A_10, covbases_A_15, covbases_A_20,\n",
        "                              covbases_C_1, covbases_C_5, covbases_C_10, covbases_C_15, covbases_C_20,\n",
        "                              covbases_G_1, covbases_G_5, covbases_G_10, covbases_G_15, covbases_G_20,\n",
        "                              covbases_T_1, covbases_T_5, covbases_T_10, covbases_T_15, covbases_T_20,\n",
        "                              coverage_A, coverage_C, coverage_G, coverage_T, num_reads_mapped_A, num_reads_mapped_C,\n",
        "                              num_reads_mapped_G, num_reads_mapped_T\n",
        "                              ))\n",
        "\n",
        "    df = pd.DataFrame(coverage_data, columns=['Region_name', 'gene_id_name', 'chrom', 'start', 'end',\n",
        "                                              'covpercent_depth_1', 'covpercent_depth_5', 'covpercent_depth_10',\n",
        "                                              'covpercent_depth_15',\n",
        "                                              'covpercent_depth_20', 'covpercent_depth_25',\n",
        "                                              'covbases_depth_1', 'covbases_depth_5', 'covbases_depth_10',\n",
        "                                              'covbases_depth_15',\n",
        "                                              'covbases_depth_20', 'covbases_depth_25',\n",
        "                                              'covpercent_depth_G_1', 'covpercent_depth_G_5', 'covpercent_depth_G_10',\n",
        "                                              'covpercent_depth_G_15', 'covpercent_depth_G_20',\n",
        "                                              'covpercent_depth_C_1', 'covpercent_depth_C_5', 'covpercent_depth_C_10',\n",
        "                                              'covpercent_depth_C_15', 'covpercent_depth_C_20',\n",
        "                                              'covpercent_depth_T_1', 'covpercent_depth_T_5', 'covpercent_depth_T_10',\n",
        "                                              'covpercent_depth_T_15', 'covpercent_depth_T_20',\n",
        "                                              'covpercent_depth_A_1', 'covpercent_depth_A_5', 'covpercent_depth_A_10',\n",
        "                                              'covpercent_depth_A_15', 'covpercent_depth_A_20',\n",
        "                                              'covbases_depth_A_1', 'covbases_depth_A_5', 'covbases_depth_A_10',\n",
        "                                              'covbases_depth_A_15', 'covbases_depth_A_20',\n",
        "                                              'covbases_depth_C_1', 'covbases_depth_C_5', 'covbases_depth_C_10',\n",
        "                                              'covbases_depth_C_15', 'covbases_depth_C_20',\n",
        "                                              'covbases_depth_G_1', 'covbases_depth_G_5', 'covbases_depth_G_10',\n",
        "                                              'covbases_depth_G_15',\n",
        "                                              'covbases_depth_G_20',\n",
        "                                              'covbases_depth_T_1', 'covbases_depth_T_5', 'covbases_depth_T_10',\n",
        "                                              'covbases_depth_T_15',\n",
        "                                              'covbases_depth_T_20',\n",
        "                                              'coverage_A_all', 'coverage_C_all', 'coverage_G_all', 'coverage_T_all',\n",
        "                                              'num_reads_mapped_A', 'num_reads_mapped_C', 'num_reads_mapped_G',\n",
        "                                              'num_reads_mapped_T'\n",
        "                                              ])\n",
        "    bam.close()\n",
        "\n",
        "    print(\"Gene coverage calculation completed.\")\n",
        "    df.to_csv(bam_file.split('.')[0] + 'updated_gene_wise_coverage.csv')\n",
        "\n",
        "    return df\n",
        "\n",
        "directory = \"/lustre/priti.devi/abuzar/DR/BAM/DR\"\n",
        "bam_files = [file_name for file_name in os.listdir(directory) if file_name.endswith(\".bam\")]\n",
        "for bam_file in bam_files:\n",
        "    bam_path = os.path.join(directory, bam_file)\n",
        "\n",
        "    # Call the function x with the BAM and GTF file paths\n",
        "    result_df = calculate_gene_coverage(bam_file, df)"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "nWCwdNIV6bW7"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}