{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "**Install and Import necessary libraries**\n",
        "\n",
        "\n",
        "1.   pysam - python wrapper of samtool\n",
        "2.   samtool - need this to index the bam file\n",
        "3.   pandas  \n",
        "4.   numpy\n",
        "5.   Array \n",
        "6.   subprocess - to execute the samtool command in \n",
        "     python environment\n",
        "\n"
      ],
      "metadata": {
        "id": "9TWwEPUsG1Pq"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "id": "beYKfUxdRN72"
      },
      "outputs": [],
      "source": [
        "#uncomment this and run \n",
        "#!pip install pysam\n",
        "#!apt-get install samtools"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Here we define a function named **calculate_gene_coverage** which takes two input bam file and gtf file.\n",
        "\n",
        "# **Step by step execuition of this function**\n",
        "\n",
        "\n",
        "1.   Index the bam file using \n",
        "     samtool via subprocessing.\n",
        "2.   Read the bam file in read and \n",
        "     binary format and align the bam file.\n",
        "3.   Process the gtf file to get the coordinates of    \n",
        "     the gene and its name.\n",
        "4.   Initiate an empty variable named coverage_data so \n",
        "     as to append all the output of the algorithms here.\n",
        "5.   Iterate over all the rows of the gtf file for \n",
        "     Gene name , start position and end position.\n",
        "6.   count_coverage method is used in the bam object. \n",
        "     this count_coverage method return 4 array.array \n",
        "     list consists of the depth of each nucleotide(ATGC) at each position. like if our ending and starting position will be like this 5 and 2 respectively then the length of each array should be 5 - 2 that is 3 and each 3 position.\n",
        "     This 3 position will be mapped by the reads \n",
        "     and algorithms return the counts of the reads mapped at each position with the respective bases lets suppose. \n",
        "     array will be like that( https://pysam.readthedocs.io/en/latest/api.html)\n",
        "     A([3, 0, 3])\n",
        "     T([5, 1, 0])\n",
        "     G([3, 2, 9])\n",
        "     C([7, 0, 0])\n",
        "\n",
        "     These are the depth of each base. \n",
        "\n",
        "  8. To get the covered percentage we need the covered \n",
        "     bases like how many bases are covered.\n",
        "     as we see in the above example.\n",
        "     we need those depth which is equals to or greater than 1.\n",
        "     in the above example we have 8 covered bases.\n",
        "     so the coverage percentage should be \n",
        "      **covbase/(end - start + 1) * 100**\n",
        "\n",
        "  9. In the end this function return a dataframe which consists of gene name, region name, start and end position, covbases and coverage percentage\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "-EPRhneEHnSl"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#result_df"
      ],
      "metadata": {
        "id": "752OLNS049G6"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Mount the google drive to ur colab and locate your respective file here and run the below given code without any changing.\n",
        "\n",
        "Just need to change here bam_file = \"/path/to/your_file.bam\" #path to bam file\n",
        "gtf_file = \"/path/to/your_file.gtf\" #path to gtf file"
      ],
      "metadata": {
        "id": "vcsQOPYbNdhC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import subprocess\n",
        "import pysam\n",
        "import pandas as pd\n",
        "\n",
        "def calculate_gene_coverage(bam_file, gtf_file):\n",
        "    subprocess.run([\"samtools\", \"index\", bam_file])\n",
        "    bam = pysam.AlignmentFile(bam_file, \"rb\")\n",
        "    gene_annotations = pd.read_csv(gtf_file, sep='\\t', comment='#', header=None, dtype={0: str})\n",
        "    gene_annotations = gene_annotations[gene_annotations[2] == \"gene\"]\n",
        "    coverage_data = []\n",
        "    for _, gene in gene_annotations.iterrows():\n",
        "        chrom = gene[0]\n",
        "        start = gene[3]\n",
        "        end = gene[4]\n",
        "        coverage = bam.count_coverage(chrom, start, end)\n",
        "        covbases = sum(sum(depth >= 1 for depth in depth_array) for depth_array in coverage)\n",
        "        covpercent = (covbases / (end - start + 1)) * 100\n",
        "        gene_name = gene[8].split(\";\")[0].split(\" \")[1].replace('\"', '')\n",
        "        coverage_data.append((gene_name, chrom, start, end, covbases, covpercent))\n",
        "    df = pd.DataFrame(coverage_data, columns=['Gene', 'Chromosome', 'Start', 'End', 'CovBases', 'CovPercent'])\n",
        "    bam.close()\n",
        "    return df\n",
        "\n",
        "#please specify the correct file location after mounting change the path below\n",
        "\n",
        "bam_file = \"/content/drive/MyDrive/gene wise coverage/IGIB113050041V.bam\" #path to bam file\n",
        "gtf_file = \"/content/drive/MyDrive/gene wise coverage/Homo_sapiens.GRCh38.106.gtf\" #path to gtf file\n",
        "\n",
        "result_df_ = calculate_gene_coverage(bam_file, gtf_file)\n"
      ],
      "metadata": {
        "id": "KpCbxZRsB_Sh"
      },
      "execution_count": 7,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "result_df_"
      ],
      "metadata": {
        "id": "R888C1onCpp1",
        "outputId": "0199aeee-aa6d-4f30-a9d7-9b7cbaba6969",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        }
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                  Gene  Chromosome    Start      End  CovBases  CovPercent\n",
              "0      ENSG00000186827           1  1211340  1214153       174    6.183369\n",
              "1      ENSG00000186891           1  1203508  1206592        18    0.583468\n",
              "2      ENSG00000160072           1  1471765  1497848      4664   17.880693\n",
              "3      ENSG00000260179           1  1249777  1251334       318   20.410783\n",
              "4      ENSG00000234396           1  2212523  2220738       843   10.260467\n",
              "...                ...         ...      ...      ...       ...         ...\n",
              "61547  ENSG00000271254  KI270711.1     4612    29626      6900   27.583450\n",
              "61548  ENSG00000275987  KI270713.1    30437    30580         0    0.000000\n",
              "61549  ENSG00000268674  KI270713.1    35407    35916       154   30.196078\n",
              "61550  ENSG00000277475  KI270713.1    31698    32528       175   21.058965\n",
              "61551  ENSG00000275405  KI270713.1    21861    22024       163   99.390244\n",
              "\n",
              "[61552 rows x 6 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-c790250b-fa6d-4497-a3d8-50a88e314856\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Gene</th>\n",
              "      <th>Chromosome</th>\n",
              "      <th>Start</th>\n",
              "      <th>End</th>\n",
              "      <th>CovBases</th>\n",
              "      <th>CovPercent</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ENSG00000186827</td>\n",
              "      <td>1</td>\n",
              "      <td>1211340</td>\n",
              "      <td>1214153</td>\n",
              "      <td>174</td>\n",
              "      <td>6.183369</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ENSG00000186891</td>\n",
              "      <td>1</td>\n",
              "      <td>1203508</td>\n",
              "      <td>1206592</td>\n",
              "      <td>18</td>\n",
              "      <td>0.583468</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ENSG00000160072</td>\n",
              "      <td>1</td>\n",
              "      <td>1471765</td>\n",
              "      <td>1497848</td>\n",
              "      <td>4664</td>\n",
              "      <td>17.880693</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ENSG00000260179</td>\n",
              "      <td>1</td>\n",
              "      <td>1249777</td>\n",
              "      <td>1251334</td>\n",
              "      <td>318</td>\n",
              "      <td>20.410783</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>ENSG00000234396</td>\n",
              "      <td>1</td>\n",
              "      <td>2212523</td>\n",
              "      <td>2220738</td>\n",
              "      <td>843</td>\n",
              "      <td>10.260467</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>61547</th>\n",
              "      <td>ENSG00000271254</td>\n",
              "      <td>KI270711.1</td>\n",
              "      <td>4612</td>\n",
              "      <td>29626</td>\n",
              "      <td>6900</td>\n",
              "      <td>27.583450</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>61548</th>\n",
              "      <td>ENSG00000275987</td>\n",
              "      <td>KI270713.1</td>\n",
              "      <td>30437</td>\n",
              "      <td>30580</td>\n",
              "      <td>0</td>\n",
              "      <td>0.000000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>61549</th>\n",
              "      <td>ENSG00000268674</td>\n",
              "      <td>KI270713.1</td>\n",
              "      <td>35407</td>\n",
              "      <td>35916</td>\n",
              "      <td>154</td>\n",
              "      <td>30.196078</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>61550</th>\n",
              "      <td>ENSG00000277475</td>\n",
              "      <td>KI270713.1</td>\n",
              "      <td>31698</td>\n",
              "      <td>32528</td>\n",
              "      <td>175</td>\n",
              "      <td>21.058965</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>61551</th>\n",
              "      <td>ENSG00000275405</td>\n",
              "      <td>KI270713.1</td>\n",
              "      <td>21861</td>\n",
              "      <td>22024</td>\n",
              "      <td>163</td>\n",
              "      <td>99.390244</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>61552 rows × 6 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-c790250b-fa6d-4497-a3d8-50a88e314856')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-c790250b-fa6d-4497-a3d8-50a88e314856 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-c790250b-fa6d-4497-a3d8-50a88e314856');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#import subprocess\n",
        "#import pysam\n",
        "#import pandas as pd\n",
        "#from array import array\n",
        "\n",
        "#def calculate_gene_coverage(bam_file, gtf_file):\n",
        " #   subprocess.run([\"samtools\", \"index\", bam_file])\n",
        "  #  bam = pysam.AlignmentFile(bam_file, \"rb\")\n",
        "   # gene_annotations = pd.read_csv(gtf_file, sep='\\t', comment='#', header=None, dtype={0: str})\n",
        "    #gene_annotations = gene_annotations[gene_annotations[2] == \"gene\"]\n",
        "    #coverage_data = []\n",
        "    #for _, gene in gene_annotations.iterrows():\n",
        "     #   chrom = gene[0]\n",
        "      #  start = gene[3]\n",
        "       # end = gene[4]\n",
        "       # coverage = bam.count_coverage(chrom,start,end)\n",
        "        #A = pd.Series(coverage[0])\n",
        "        #C = pd.Series(coverage[1])\n",
        "        #G = pd.Series(coverage[2])\n",
        "        #T = pd.Series(coverage[3])\n",
        "        #A = pd.DataFrame(A, columns=['A'])\n",
        "        #T = pd.DataFrame(T, columns=['T'])\n",
        "        #G = pd.DataFrame(G, columns=['G'])\n",
        "        #C = pd.DataFrame(C, columns=['C'])\n",
        "        #A = pd.DataFrame(A) \n",
        "        #C = pd.DataFrame(C)\n",
        "        #G = pd.DataFrame(G)\n",
        "        #T = pd.DataFrame(T)\n",
        "        #depthA = A[A['A'] >=1]\n",
        "        #depthT = T[T['T'] >=1]\n",
        "        #depthG = G[G['G'] >=1]\n",
        "        #depthC = C[C['C'] >=1]\n",
        "        #covbases = depthA.shape[0] + depthT.shape[0] + depthG.shape[0] + depthC.shape[0]\n",
        "        #covpercent = (covbases / (end - start + 1)) * 100 \n",
        "        #gene_name = gene[8].split(\";\")[0].split(\" \")[1].replace('\"', '')\n",
        "        #coverage_data.append((gene_name, chrom, start, end, covbases, covpercent))\n",
        "    #df = pd.DataFrame(coverage_data, columns=['Gene', 'Chromosome', 'Start', 'End', 'CovBases', 'CovPercent'])\n",
        "\n",
        "    #bam.close()\n",
        "\n",
        "    #return df\n",
        "\n",
        "#bam_file = \"/content/drive/MyDrive/gene wise coverage/IGIB113050041V.bam\"\n",
        "#gtf_file = \"/content/drive/MyDrive/gene wise coverage/Homo_sapiens.GRCh38.106.gtf\"\n",
        "\n",
        "#result_df = calculate_gene_coverage(bam_file, gtf_file)\n",
        "\n"
      ],
      "metadata": {
        "id": "EoVKROvjcSB7"
      },
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#result_df.to_csv('IGIB113050041V_gene_wise_coverage.csv')"
      ],
      "metadata": {
        "id": "3r_DZQYbe0lo"
      },
      "execution_count": 10,
      "outputs": []
    }
  ]
}